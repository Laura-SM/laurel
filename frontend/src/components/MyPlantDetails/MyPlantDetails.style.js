import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  centralContainer: {
    alignItems: 'center',
  },
  image: {
    width: 290,
    height: 290,
    borderRadius: 10,
    margin: 20,
    backgroundColor: '#E5E5E5',
  },
  featuresContainer: {
    width: 300,
    alignItems: 'flex-start',
  },
});

export default styles;
