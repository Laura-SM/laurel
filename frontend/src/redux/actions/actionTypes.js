export default {
  LOAD_PLANTS: 'LOAD_PLANTS',
  ADD_PLANT: 'ADD_PLANT',
  UPDATE_PLANT: 'UPDATE_PLANT',
  DELETE_PLANT: 'DELETE_PLANT',
  LOAD_PLANT: 'LOAD_PLANT',
  PLANTS_ERROR: 'PLANTS_ERROR',

  GET_USER: 'GET_USER',
  USER_ERROR: 'USER_ERROR',
};
