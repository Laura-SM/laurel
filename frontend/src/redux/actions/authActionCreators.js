import axios from 'axios';
import actionTypes from './actionTypes';
import {env} from '../../../.env.js';

const url = env.REACT_APP_PORT;

export function signInUser(user) {
  return async dispatch => {
    try {
      const {data} = await axios.post(`${url}/signin`, user);
      dispatch({
        type: actionTypes.SIGNIN_USER,
        userTokens: data,
      });
    } catch (error) {
      dispatch({
        type: actionTypes.USER_ERROR,
      });
    }
  };
}

export function signUpUser(user) {
  return async dispatch => {
    try {
      const {data} = await axios.post(`${url}/signup`, user);
      dispatch({
        type: actionTypes.SIGNUP_USER,
        user: data,
      });
    } catch (error) {
      dispatch({
        type: actionTypes.USER_ERROR,
      });
    }
  };
}
