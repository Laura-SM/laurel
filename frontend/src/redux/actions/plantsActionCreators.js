import axios from 'axios';
import actionTypes from './actionTypes';
import {env} from '../../../.env.js';

const url = env.REACT_APP_PORT;

export function loadPlants() {
  return async dispatch => {
    try {
      const {data} = await axios(`${url}/plants`);
      dispatch({
        type: actionTypes.LOAD_PLANTS,
        plants: data,
      });
    } catch (error) {
      dispatch({
        type: 'PLANTS_ERROR',
      });
    }
  };
}

export function addPlant(plant) {
  return async dispatch => {
    try {
      const {data} = await axios.post(`${url}/plants`, plant);
      dispatch({
        type: actionTypes.ADD_PLANT,
        plant: data,
      });
    } catch (error) {
      dispatch({
        type: 'PLANTS_ERROR',
      });
    }
  };
}

export function deletePlant(plantId) {
  return async dispatch => {
    try {
      await axios.delete(`${url}/plants/${plantId}`);
      dispatch({
        type: actionTypes.DELETE_PLANT,
        plantId,
      });
    } catch (error) {
      dispatch({
        type: 'PLANTS_ERROR',
      });
    }
  };
}

export function updatePlant(plant) {
  return async dispatch => {
    try {
      const {data} = await axios.put(`${url}/plants/${plant._id}`, plant);
      dispatch({
        type: actionTypes.UPDATE_PLANT,
        plant: data,
      });
    } catch (error) {
      dispatch({
        type: 'PLANTS_ERROR',
      });
    }
  };
}

export function loadPlant(plantId) {
  return async dispatch => {
    try {
      const {data} = await axios(`${url}/plants/${plantId}`);
      dispatch({
        type: actionTypes.LOAD_PLANT,
        selectedPlant: data,
      });
    } catch (error) {
      dispatch({
        type: 'PLANTS_ERROR',
      });
    }
  };
}
