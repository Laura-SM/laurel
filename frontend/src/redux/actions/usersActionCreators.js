import axios from 'axios';
import actionTypes from './actionTypes';

const url = 'http://192.168.8.104:2021/login';

export function loadUser(user) {
  return async dispatch => {
    try {
      const {data} = await axios.post(url, user);
      dispatch({
        type: actionTypes.SIGNIN_USER,
        userTokens: data,
      });
    } catch (error) {
      dispatch({
        type: actionTypes.USER_ERROR,
      });
    }
  };
}
