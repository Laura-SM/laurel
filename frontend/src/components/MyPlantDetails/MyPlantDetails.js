/* eslint-disable react-hooks/exhaustive-deps */
import React, {useEffect} from 'react';
import {Text, View, Image, TouchableOpacity} from 'react-native';
import {connect} from 'react-redux';
import {deletePlant, loadPlant} from '../../redux/actions/plantsActionCreators';
import {updateUser} from '../../redux/actions/usersActionCreators';
import {useNavigation} from '@react-navigation/native';
import styles from './MyPlantDetails.style';
import globalStyles from '../../styles/global.styles';

const MyPlantDetails = ({
  userAccess,
  selectedPlant,
  dispatch,
  route,
  navigation: {goBack},
}) => {
  const {plantId} = route.params;
  const navigation = useNavigation();
  let myPlantsIds = userAccess.user.plants;
  const onPressDeletePlant = () => {
    userAccess.user.plants = myPlantsIds.filter(id => id !== plantId);
    dispatch(updateUser(userAccess.user));
    dispatch(deletePlant(plantId));
    navigation.navigate('MyPlants', {myPlantsIds: myPlantsIds});
  };

  useEffect(() => {
    dispatch(loadPlant(plantId));
  }, [route.params]);

  return (
    <View style={globalStyles.mainContainer}>
      <View style={globalStyles.headerContainer}>
        <Text style={globalStyles.titleText}>{selectedPlant.name}</Text>
        <TouchableOpacity
          style={globalStyles.roundButton}
          onPress={() => onPressDeletePlant()}>
          <Image source={require('../../icons/delete24.png')} />
        </TouchableOpacity>
      </View>
      <View style={styles.centralContainer}>
        <Image style={styles.image} source={{uri: selectedPlant.image}} />
        <View style={styles.featuresContainer}>
          <View style={styles.iconsContainer}>
            <Image source={require('../../icons/room24.png')} />
            <Text style={globalStyles.text}>{selectedPlant.room}</Text>
          </View>
          <View style={styles.iconsContainer}>
            <Image source={require('../../icons/water24.png')} />
            <Text style={globalStyles.text}>{selectedPlant.nextWaterDate}</Text>
          </View>
          <View style={styles.iconsContainer}>
            <Image source={require('../../icons/mist24.png')} />
            <Text style={globalStyles.text}>{selectedPlant.nextMistDate}</Text>
          </View>
          <View style={styles.iconsContainer}>
            <Image source={require('../../icons/settings24.png')} />
            <Text style={globalStyles.text}>
              {selectedPlant.nextTransplantDate}
            </Text>
          </View>
        </View>
      </View>
      <View style={globalStyles.bottomContainer}>
        <TouchableOpacity
          style={globalStyles.roundButton}
          onPress={() => goBack()}>
          <Image source={require('../../icons/goBack24.png')} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

function mapStateToProps({selectedPlant, userAccess}) {
  return {
    selectedPlant,
    userAccess,
  };
}

export default connect(mapStateToProps)(MyPlantDetails);
