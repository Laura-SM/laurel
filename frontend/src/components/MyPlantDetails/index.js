/* eslint-disable react-hooks/exhaustive-deps */
import React, {useEffect} from 'react';
import {Text, View, Image, TouchableOpacity} from 'react-native';
import {connect} from 'react-redux';
import {deletePlant, loadPlant} from '../../redux/actions/plantsActionCreators';
import {updateUser} from '../../redux/actions/usersActionCreators';
import styles from './styles';
import globalStyles from '../../styles/globalStyles';

const MyPlantDetails = ({
  userAccess,
  selectedPlant,
  dispatch,
  route,
  navigation,
}) => {
  const {plantId} = route.params;

  const waterDate = new Date(selectedPlant.nextWaterDate).toLocaleDateString();
  const mistDate = new Date(selectedPlant.nextMistDate).toLocaleDateString();
  const transplantDate = new Date(
    selectedPlant.nextTransplantDate,
  ).toLocaleDateString();

  let myPlantsIds = userAccess.user.plants;

  const onPressDeletePlant = () => {
    userAccess.user.plants = myPlantsIds.filter(id => id !== plantId);
    dispatch(updateUser(userAccess.user));
    dispatch(deletePlant(plantId));
    navigation.navigate('MyPlants', {myPlantsIds: myPlantsIds});
  };

  useEffect(() => {
    dispatch(loadPlant(plantId));
  }, [route.params]);

  return (
    <View style={globalStyles.mainContainer}>
      <View style={globalStyles.headerContainer}>
        <Text style={globalStyles.titleText}>{selectedPlant.name}</Text>
        <TouchableOpacity
          testID="deletePlantButton"
          style={globalStyles.roundButton}
          onPress={() => onPressDeletePlant()}>
          <Image source={require('../../icons/delete24.png')} />
        </TouchableOpacity>
      </View>
      <View style={styles.centralContainer}>
        <Image style={styles.image} source={{uri: selectedPlant.image}} />
        <View style={styles.featuresContainer}>
          <View style={globalStyles.iconsContainer}>
            <Image source={require('../../icons/room24.png')} />
            <Text style={globalStyles.text}>{selectedPlant.room}</Text>
          </View>
          <View style={globalStyles.iconsContainer}>
            <Image source={require('../../icons/water24.png')} />
            <Text style={globalStyles.text}>{waterDate}</Text>
          </View>
          <View style={globalStyles.iconsContainer}>
            <Image source={require('../../icons/mist24.png')} />
            <Text style={globalStyles.text}>{mistDate}</Text>
          </View>
          <View style={globalStyles.iconsContainer}>
            <Image source={require('../../icons/settings24.png')} />
            <Text style={globalStyles.text}>{transplantDate}</Text>
          </View>
        </View>
      </View>
      <View style={globalStyles.bottomContainer}>
        <TouchableOpacity
          testID="goBackButton"
          style={globalStyles.roundButton}
          onPress={() => navigation.goBack()}>
          <Image source={require('../../icons/goBack24.png')} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

function mapStateToProps({selectedPlant, userAccess}) {
  return {
    selectedPlant,
    userAccess,
  };
}

export default connect(mapStateToProps)(MyPlantDetails);
