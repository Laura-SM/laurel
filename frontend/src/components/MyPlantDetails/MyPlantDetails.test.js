import 'react-native';
import React from 'react';
import MyPlantDetails from './index';
import * as userActions from '../../redux/actions/usersActionCreators';
import * as actions from '../../redux/actions/plantsActionCreators';
import thunk from 'redux-thunk';
import Store from 'redux-mock-store';
import {Provider} from 'react-redux';
import {render, fireEvent} from '@testing-library/react-native';

describe('Given a MyPlantDetails component', () => {
  const mockStore = Store([thunk]);
  let navigation;
  let myStore;
  let route = {
    params: {plantId: '1a'},
  };

  beforeEach(() => {
    jest.spyOn(userActions, 'updateUser').mockReturnValueOnce({type: ''});
    myStore = mockStore({
      userAccess: {
        user: {email: 'laura@email.com', password: '123', plants: ['1a']},
      },
      selectedPlant: {
        name: 'kentia',
        _id: '1a',
        image: 'img.com',
        room: 'bedroom',
      },
    });
    navigation = {
      navigate: jest.fn(),
      goBack: jest.fn(),
    };
  });

  describe('When invoques MyPlantDetails function', () => {
    test('Should render a plant with detail', () => {
      const detail = render(
        <Provider store={myStore}>
          <MyPlantDetails route={route} />
        </Provider>,
      );
      expect(detail).toMatchSnapshot();
    });
  });

  describe('When goBack button is pressed', () => {
    test('Then navigation.goBack is invoked', () => {
      const {getByTestId} = render(
        <Provider store={myStore}>
          <MyPlantDetails route={route} navigation={navigation} />
        </Provider>,
      );
      const goBack = getByTestId('goBackButton');
      fireEvent.press(goBack);

      expect(navigation.goBack).toHaveBeenCalled();
    });
  });

  describe('When deletePlant button is pressed', () => {
    test('Then updateUser is invoked', () => {
      const {getByTestId} = render(
        <Provider store={myStore}>
          <MyPlantDetails route={route} navigation={navigation} />
        </Provider>,
      );
      const deletePlant = getByTestId('deletePlantButton');
      fireEvent.press(deletePlant);

      expect(userActions.updateUser).toHaveBeenCalled();
    });
  });
});
