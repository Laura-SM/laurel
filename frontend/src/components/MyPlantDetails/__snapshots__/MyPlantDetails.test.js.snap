// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MyPlantDetails component When invoques MyPlantDetails function Should render a plant with detail 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 80,
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "margin": 20,
        }
      }
    >
      kentia
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#B8E0D2",
          "borderRadius": 50,
          "height": 60,
          "justifyContent": "center",
          "margin": 20,
          "opacity": 1,
          "width": 60,
        }
      }
      testID="deletePlantButton"
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/icons/delete24.png",
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "img.com",
        }
      }
      style={
        Object {
          "backgroundColor": "#E5E5E5",
          "borderRadius": 10,
          "height": 290,
          "margin": 20,
          "width": 290,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "width": 300,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginVertical": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/icons/room24.png",
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginHorizontal": 20,
            }
          }
        >
          bedroom
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginVertical": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/icons/water24.png",
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginHorizontal": 20,
            }
          }
        >
          Invalid Date
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginVertical": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/icons/mist24.png",
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginHorizontal": 20,
            }
          }
        >
          Invalid Date
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginVertical": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/icons/settings24.png",
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginHorizontal": 20,
            }
          }
        >
          Invalid Date
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#B8E0D2",
          "borderRadius": 50,
          "height": 60,
          "justifyContent": "center",
          "margin": 20,
          "opacity": 1,
          "width": 60,
        }
      }
      testID="goBackButton"
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/icons/goBack24.png",
          }
        }
      />
    </View>
  </View>
</View>
`;
